File: ui/e2e/mocks/handlers/virtual-keys.ts
import type { Page } from "@playwright/test";
import type { MockStore } from "../store";
import { json, error } from "./utils";

// Mirrors the REAL Go /api/virtual-keys CRUD (internal/admin/virtualkeys.go, plan
// §1.4 / §1.6 / §8 ESC-2). Response bodies use the Go envelope shapes:
//   GET  /api/virtual-keys -> {data:{virtual_keys:[virtualKeyDTO]}}
//   POST /api/virtual-keys -> {data:{virtual_key:virtualKeyDTO}}
//   GET/PUT /api/virtual-keys/{id} -> {data:{virtual_key:virtualKeyDTO}}
//   DELETE /api/virtual-keys/{id} -> {data:{message}}
// Create/update REQUIRE provider_configs with a non-empty key_ids per config
// (virtualkeys.go:55-66) — the KeyIDs editor (plan §1.6) writes provider_configs.
interface ProviderConfigLike {
  provider?: string;
  allowed_models?: string[];
  key_ids?: string[];
}

function validateProviderConfigs(body: { provider_configs?: ProviderConfigLike[] }): string | null {
  const configs = body.provider_configs;
  if (!configs || configs.length === 0) return "provider_configs is required";
  for (let i = 0; i < configs.length; i++) {
    const pc = configs[i];
    if (!pc.provider) return `provider_configs[${i}].provider is required`;
    if (!pc.allowed_models || pc.allowed_models.length === 0) return `provider_configs[${i}].allowed_models is required`;
    if (!pc.key_ids || pc.key_ids.length === 0) return `provider_configs[${i}].key_ids is required`;
  }
  return null;
}

export function registerVirtualKeysHandlers(page: Page, store: MockStore) {
  page.route("/api/virtual-keys", async (route) => {
    const method = route.request().method();
    if (method === "GET") {
      return json(route, { virtual_keys: Array.from(store.virtualKeys.values()) });
    }
    if (method === "POST") {
      const body = await route.request().postDataJSON();
      if (!body.name) return error(route, "name is required", 400);
      const invalid = validateProviderConfigs(body);
      if (invalid) return error(route, invalid, 400);
      const id = store.nextId();
      const now = Math.floor(Date.now() / 1000);
      const vk = {
        id,
        key: `vk-${Math.random().toString(36).slice(2, 12)}`,
        name: body.name,
        provider_configs: body.provider_configs,
        budget: body.budget,
        rate_limit_rpm: body.rate_limit_rpm,
        is_active: body.is_active ?? true,
        created_at: now,
        updated_at: now,
      };
      store.virtualKeys.set(id, vk as never);
      return json(route, { virtual_key: vk });
    }
    return route.continue();
  });
  page.route(/\/api\/virtual-keys\/[^/]+$/, async (route) => {
    const id = route.request().url().split("/").pop()!;
    const method = route.request().method();
    if (method === "GET") {
      const vk = store.virtualKeys.get(id);
      return vk ? json(route, { virtual_key: vk }) : error(route, "virtual key not found", 404);
    }
    if (method === "PUT") {
      const body = await route.request().postDataJSON();
      const existing = store.virtualKeys.get(id);
      if (!existing) return error(route, "virtual key not found", 404);
      if (!body.name) return error(route, "name is required", 400);
      const invalid = validateProviderConfigs(body);
      if (invalid) return error(route, invalid, 400);
      const updated = {
        ...existing,
        name: body.name,
        provider_configs: body.provider_configs,
        budget: body.budget,
        rate_limit_rpm: body.rate_limit_rpm,
        is_active: body.is_active ?? (existing as { is_active?: boolean }).is_active ?? true,
        updated_at: Math.floor(Date.now() / 1000),
      };
      store.virtualKeys.set(id, updated as never);
      return json(route, { virtual_key: updated });
    }
    if (method === "DELETE") {
      store.virtualKeys.delete(id);
      return json(route, { message: "Virtual key deleted successfully" });
    }
    return route.continue();
  });
}

File: internal/admin/virtualkeys.go
package admin

import (
	"encoding/json"
	"errors"
	"fmt"

	"github.com/bloodf/g0router/internal/schemas"
	"github.com/bloodf/g0router/internal/store"
	"github.com/valyala/fasthttp"
)

type virtualKeyDTO struct {
	ID              string                  `json:"id"`
	Key             string                  `json:"key"`
	Name            string                  `json:"name"`
	ProviderConfigs []schemas.ProviderConfig `json:"provider_configs"`
	Budget          *schemas.Budget         `json:"budget,omitempty"`
	RateLimitRPM    *int                    `json:"rate_limit_rpm,omitempty"`
	IsActive        bool                    `json:"is_active"`
	CreatedAt       int64                   `json:"created_at"`
	UpdatedAt       int64                   `json:"updated_at"`
}

func toVirtualKeyDTO(vk *store.VirtualKey) virtualKeyDTO {
	dto := virtualKeyDTO{
		ID:              vk.ID,
		Key:             vk.Key,
		Name:            vk.Name,
		ProviderConfigs: vk.ProviderConfigs,
		Budget:          vk.Budget,
		RateLimitRPM:    vk.RateLimitRPM,
		IsActive:        vk.IsActive,
		CreatedAt:       vk.CreatedAt,
		UpdatedAt:       vk.UpdatedAt,
	}
	if dto.ProviderConfigs == nil {
		dto.ProviderConfigs = []schemas.ProviderConfig{}
	}
	return dto
}

type virtualKeyRequest struct {
	Name            string                  `json:"name"`
	ProviderConfigs []schemas.ProviderConfig `json:"provider_configs"`
	Budget          *schemas.Budget         `json:"budget,omitempty"`
	RateLimitRPM    *int                    `json:"rate_limit_rpm,omitempty"`
	IsActive        *bool                   `json:"is_active,omitempty"`
}

func validateVirtualKeyRequest(req *virtualKeyRequest) error {
	if req.Name == "" {
		return fmt.Errorf("name is required")
	}
	if len(req.ProviderConfigs) == 0 {
		return fmt.Errorf("provider_configs is required")
	}
	for i, pc := range req.ProviderConfigs {
		if pc.Provider == "" {
			return fmt.Errorf("provider_configs[%d].provider is required", i)
		}
		if len(pc.AllowedModels) == 0 {
			return fmt.Errorf("provider_configs[%d].allowed_models is required", i)
		}
		if len(pc.KeyIDs) == 0 {
			return fmt.Errorf("provider_configs[%d].key_ids is required", i)
		}
	}
	if req.Budget != nil {
		if req.Budget.Limit < 0 {
			return fmt.Errorf("budget.limit must be non-negative")
		}
		if req.Budget.Used < 0 {
			return fmt.Errorf("budget.used must be non-negative")
		}
		if req.Budget.Period != "" && req.Budget.Period != "daily" && req.Budget.Period != "weekly" && req.Budget.Period != "monthly" {
			return fmt.Errorf("budget.period must be daily, weekly, or monthly")
		}
	}
	if req.RateLimitRPM != nil && *req.RateLimitRPM < 0 {
		return fmt.Errorf("rate_limit_rpm must be non-negative")
	}
	return nil
}

// ListVirtualKeys handles GET /api/virtual-keys.
func (h *Handlers) ListVirtualKeys(ctx *fasthttp.RequestCtx) {
	vks, err := h.store.ListVirtualKeys()
	if err != nil {
		writeError(ctx, fasthttp.StatusInternalServerError, "list virtual keys")
		return
	}
	out := make([]virtualKeyDTO, 0, len(vks))
	for _, vk := range vks {
		out = append(out, toVirtualKeyDTO(vk))
	}
	writeData(ctx, fasthttp.StatusOK, map[string]any{"virtual_keys": out})
}

// CreateVirtualKey handles POST /api/virtual-keys.
func (h *Handlers) CreateVirtualKey(ctx *fasthttp.RequestCtx) {
	var req virtualKeyRequest
	if err := json.Unmarshal(ctx.PostBody(), &req); err != nil {
		writeError(ctx, fasthttp.StatusBadRequest, "invalid JSON body")
		return
	}
	if err := validateVirtualKeyRequest(&req); err != nil {
		writeError(ctx, fasthttp.StatusBadRequest, err.Error())
		return
	}

	vk := &store.VirtualKey{
		VirtualKey: schemas.VirtualKey{
			Name:            req.Name,
			ProviderConfigs: req.ProviderConfigs,
			Budget:          req.Budget,
			RateLimitRPM:    req.RateLimitRPM,
		},
	}
	created, err := h.store.CreateVirtualKey(vk)
	if err != nil {
		writeError(ctx, fasthttp.StatusInternalServerError, "create virtual key")
		return
	}
	writeData(ctx, fasthttp.StatusCreated, map[string]any{"virtual_key": toVirtualKeyDTO(created)})
}

// GetVirtualKey handles GET /api/virtual-keys/{id}.
func (h *Handlers) GetVirtualKey(ctx *fasthttp.RequestCtx) {
	id, ok := pathID(ctx.UserValue("id"))
	if !ok {
		writeError(ctx, fasthttp.StatusBadRequest, "invalid route parameter")
		return
	}
	vk, err := h.store.GetVirtualKeyByID(id)
	if errors.Is(err, store.ErrNotFound) {
		writeError(ctx, fasthttp.StatusNotFound, "virtual key not found")
		return
	}
	if err != nil {
		writeError(ctx, fasthttp.StatusInternalServerError, "load virtual key")
		return
	}
	writeData(ctx, fasthttp.StatusOK, map[string]any{"virtual_key": toVirtualKeyDTO(vk)})
}

// UpdateVirtualKey handles PUT /api/virtual-keys/{id}.
func (h *Handlers) UpdateVirtualKey(ctx *fasthttp.RequestCtx) {
	id, ok := pathID(ctx.UserValue("id"))
	if !ok {
		writeError(ctx, fasthttp.StatusBadRequest, "invalid route parameter")
		return
	}
	var req virtualKeyRequest
	if err := json.Unmarshal(ctx.PostBody(), &req); err != nil {
		writeError(ctx, fasthttp.StatusBadRequest, "invalid JSON body")
		return
	}
	if err := validateVirtualKeyRequest(&req); err != nil {
		writeError(ctx, fasthttp.StatusBadRequest, err.Error())
		return
	}

	existing, err := h.store.GetVirtualKeyByID(id)
	if errors.Is(err, store.ErrNotFound) {
		writeError(ctx, fasthttp.StatusNotFound, "virtual key not found")
		return
	}
	if err != nil {
		writeError(ctx, fasthttp.StatusInternalServerError, "load virtual key")
		return
	}

	isActive := existing.IsActive
	if req.IsActive != nil {
		isActive = *req.IsActive
	}
	vk := &store.VirtualKey{
		VirtualKey: schemas.VirtualKey{
			ID:              id,
			Name:            req.Name,
			ProviderConfigs: req.ProviderConfigs,
			Budget:          req.Budget,
			RateLimitRPM:    req.RateLimitRPM,
		},
		IsActive: isActive,
	}
	if err := h.store.UpdateVirtualKey(vk); err != nil {
		writeError(ctx, fasthttp.StatusInternalServerError, "update virtual key")
		return
	}

	updated, err := h.store.GetVirtualKeyByID(id)
	if err != nil {
		writeError(ctx, fasthttp.StatusInternalServerError, "load virtual key")
		return
	}
	writeData(ctx, fasthttp.StatusOK, map[string]any{"virtual_key": toVirtualKeyDTO(updated)})
}

// DeleteVirtualKey handles DELETE /api/virtual-keys/{id}.
func (h *Handlers) DeleteVirtualKey(ctx *fasthttp.RequestCtx) {
	id, ok := pathID(ctx.UserValue("id"))
	if !ok {
		writeError(ctx, fasthttp.StatusBadRequest, "invalid route parameter")
		return
	}
	if err := h.store.DeleteVirtualKey(id); errors.Is(err, store.ErrNotFound) {
		writeError(ctx, fasthttp.StatusNotFound, "virtual key not found")
		return
	} else if err != nil {
		writeError(ctx, fasthttp.StatusInternalServerError, "delete virtual key")
		return
	}
	writeData(ctx, fasthttp.StatusOK, map[string]any{"message": "Virtual key deleted successfully"})
}
